TypeScript strict mode support + docs (#5491)
redwoodjs/redwood: docs/docs/typescript/generated-types.md

:::

## What's happening under the hood

Redwood uses [GraphQL Code Generator](https://www.graphql-code-generator.com) to generate types for your GraphQL queries, mutations and SDLs. We configure it to use the types from your generated prisma client too on the API side, to make sure your resolvers are strongly typed.

### Customising codegen config
While the defaults are configured so that things JustWork™️, you can customize them by adding a `./codegen.yml` file to the root of your project. Your custom settings will be merged with the built-in ones.

:::info
If you're curious about the built-in settings, they can be found [here](https://github.com/redwoodjs/redwood/blob/main/packages/internal/src/generate/graphqlCodeGen.ts) in the Redwood source. Look for the `generateTypeDefGraphQLWeb` and `generateTypeDefGraphQLApi` functions.
:::

For example, adding this `codegen.yml` to the root of your project will transform all the generated types to UPPERCASE:

```yml
# ./codegen.yml

config:
  namingConvention:

>     typeNames: change-case-all#upperCase

Author: Yeah this is the correct syntax. No idea about how yaml handles commets, maybe its just when its with a space?

---
description: A look at automatic type generation in Redwood
---

# Generated Types

To add to the TypeScript (and JavaScript!) experience, Redwood generates types for you.
These generated types not only include your GraphQL operations, but also your named routes, Cells, scenarios, and tests.

When you run `yarn rw dev`, the CLI watches files for changes and triggers type generation automatically, but you can trigger it manually too:

```
yarn rw g types
```

:::tip Getting errors trying to generate types?

If you're getting errors trying to generate types, it's worth checking the GraphQL operations in your Cells and SDLs.
Make sure that they're syntactically valid, and that every query and mutation on the web side is defined in an `*.sdl.js` file on the api side.

:::

If you're curious, you can find the generated types in the `.redwood/types`, `web/types/graphql.d.ts`, and `api/types/graphql.d.ts` directories. Broadly speaking, Redwood generates the following types:

> 1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, Cells, and layouts on the web side

Reviewer: Is there more than what's enumerated here? Maybe the "mirror" types include the api side too? For a simple project with a Post model and dbAuth setup, this is what I'm seeing:

```
.redwood/types/mirror/
├── web/
│  └── src/
│     └── pages/
│        ├── NotFoundPage/
│        │  └── index.d.ts
│        └── FatalErrorPage/
│           └── index.d.ts
└── api/
   └── src/
      ├── services/
      │  └── posts/
      │     └── index.d.ts
      └── directives/
         ├── skipAuth/
         │  └── index.d.ts
         └── requireAuth/
            └── index.d.ts
```

> 1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, Cells, and layouts on the web side

Author: ```suggestion
1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, Cells, and layouts on the web side - and services, libs, etc. on the api side
```

> 1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, Cells, and layouts on the web side - and services, libs, etc. on the api side

Reviewer: ```suggestion
1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, pages, layouts, etc. on the web side, and for your services, lib, etc. on the api side
```

2. types based on your queries and mutations on the web side (in `web/types/graphql.d.ts`)
3. types for resolvers based on your SDLs on the api side (in `api/types/graphql.d.ts`)

> 4. types for testing, `currentUser`, etc.

Reviewer: I've seen all three used and I'm a little confused if they're different: is it `currentUser`, `CurrentUser`, or CurrentUser / current user (i.e. not a code thing at all?)

Reviewer: I.e. since `CurrentUser` is capitalized, is it a type? Whereas `currentUser` is a variable somewhere? And then current user is just the concept?

Author: They're interchangable really. But `currentUser` refers to the object/property usually, `CurrentUser` refers to the type.

But it gets tricky, when you say "you can find this in your CurrentUser" - where it means both

## CurrentUser

If you've setup auth, the type for the current user on both the web and the api side gets automatically "inferred" from the `getCurrentUser` function in `api/src/lib/auth.ts`.

For example, if you specify the return type on `getCurrentUser` as...

```ts title="api/src/lib/auth.ts"
interface MyCurrentUser {
  id: string,
  roles: string[],
  email: string,
  projectId: number
}

const getCurrentUser = ({decoded}): MyCurrentUser => {
  //..
}
```

> The types for both `useAuth().currentUser` on the web side and `context.currentUser` on the api side will be the same—the `MyCurrentUser` interface.

Reviewer: I updated the copy—is this still correct?

Author: Ummm.. I think its important that we point out that the type _comes_ from the `getCurrentUser` function

Author: Actually, I covered it above. Change looks good

The import statement's specifier, `'types/graphql'`, is a [mapped path](https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping). First, TypeScript will look for the types in `web/types/graphql.d.ts`; if they're not there, it'll check `types/graphql.d.ts`. Redwood only automatically generates the former. For the latter, see [sharing types between sides](./introduction.md#sharing-types-between-sides).

But don't worry too much. If you use the generators, they template all of this for you!

## Resolver Types

Generated Services include types for query and mutation resolvers:

```ts title="api/src/services/posts.ts"
// highlight-next-line
import type { QueryResolvers, MutationResolvers } from 'types/graphql'

import { db } from 'src/lib/db'

// highlight-next-line
export const posts: QueryResolvers['posts'] = () => {
  return db.post.findMany()
}

// highlight-next-line
export const post: QueryResolvers['post'] = ({ id }) => {
  return db.post.findUnique({
    where: { id },
  })
}
```

> These types help you by making sure you're returning an object in the shape of what you've defined in your SDL. Note that these types expect you to return the _complete_ type that you've defined in your SDL. You can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.

Reviewer: I don't quite understand the last two sentences in this paragraph.

Author: What I meant to say is - "normally typing these things in graphql is a royal pain in the butt, because you can't just return a DateTime from Prisma (because graphql wants a string for a Date field).... but with Redwood, its all good, and these QueryResolver types already do this mapping for you" 

But I agree doesn't read too well.

> These types help you by making sure you're returning an object in the shape of what you've defined in your SDL. Note that these types expect you to return the _complete_ type that you've defined in your SDL. You can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.

Reviewer: ```suggestion
These types help you by making sure you're returning an object in the shape of what you've defined in your SDL. Note that these types expect you to return the _complete_ type that you've defined in your SDL. But you can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.
```

> These types help you by making sure you're returning an object in the shape of what you've defined in your SDL. Note that these types expect you to return the _complete_ type that you've defined in your SDL. You can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.

Reviewer 2: Was it here that we determined that the shape wasn't based on the SDL but rather on the Prisma schema?
```suggestion
These types help you by making sure you're returning an object in the shape of what you've defined in your Prisma schema. Note that these types expect you to return the _complete_ type that you've defined in your schema. You can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.
```

Author: Both these statements are true. If the Model exists in Prisma, it'll use the Prisma model, if it doesn't it'll use the SDL model.

Reviewer 2: Ok, can we make that clearer in the docs somehow?

Reviewer 2: Looks great now!

:::note A note on union types

Lets say that in one of your SDLs, you define a union type

```graphql
type OutOfStock {
  message: String!
}

// highlight-next-line
union CandyResult = Candy | OutOfStock

type Query {
  candy(id: String!): CandyResult @skipAuth
}
```

> These types will also be handled automatically. But if you're returning a different Prisma model, you may need to write your own resolver type, as the type generator won't know how to map the Prisma type to the GraphQL return type.

Reviewer: Just to clarify, what you're saying is you have to do this when the union is two Prisma models right?

Author: Yes that's right.

Author: So if you return something like `union CandyResult  = Candy | Chocolate` - because the typegenerator doesn't handle this usecase (I think)

> These types will also be handled automatically. But if you're returning a different Prisma model, you may need to write your own resolver type, as the type generator won't know how to map the Prisma type to the GraphQL return type.

Reviewer: ```suggestion
These types will also be handled automatically. But if you're returning a different Prisma model (instead of something like the generic `OutOfStock` type we have here, which is just a message), you may need to write your own resolver type, as the type generator won't know how to map the Prisma type to the GraphQL return type.
```

:::

## Under the Hood

Redwood uses [GraphQL Code Generator](https://www.graphql-code-generator.com) (aka graphql-codegen) to generate types for your GraphQL operations and SDLs. It's even configured to use the types from your generated Prisma Client, to make sure that your resolvers are strongly typed!

### Customizing GraphQL Code Generation

While the default settings are configured so that things just work️, you can customize them to your liking by adding a `./codegen.yml` file to the root of your project.

:::info Curious about the defaults?

You can find them [here](https://github.com/redwoodjs/redwood/blob/main/packages/internal/src/generate/graphqlCodeGen.ts) in Redwood's source. Look for the `generateTypeDefGraphQLWeb` and `generateTypeDefGraphQLApi` functions.

:::

For example, adding this `codegen.yml` to the root of your project will transform the names of the generated types to UPPERCASE:

> ```yml title="codegen.yml"

Reviewer: This one isn't a huge deal, so let's just leave it as is at the moment and I'll file an issue on docusaurus's repo or whatever does the highlighting.

config:
  namingConvention:
    typeNames: change-case-all#upperCase
```

You can configure graphql-codegen in a number of different ways: `codegen.yml`, `codegen.json`, or `codegen.js`. Even a `codegen` key in your root `package.json` will do. graphql-codegen uses [cosmiconfig](https://github.com/davidtheclark/cosmiconfig#cosmiconfig) under the hood—take a look at their docs if you want to know more.

For completeness, [here's the docs](https://www.graphql-code-generator.com/docs/config-reference/config-field) on configuring GraphQL Code Generator. Currently, Redwood only supports the root level `config` option.

:::tip Using VSCode?

As a part of type generation, the [VSCode GraphQL extension](https://marketplace.visualstudio.com/items?itemName=GraphQL.vscode-graphql) configures itself based on the merged schema Redwood generates in `.redwood/schema.graphql`.

> You can configure it further in `graphql.config.js` at the root of your project.

Reviewer: This is only related cause the extensions consumes the generated types right?

Author: Ummm the extension consumes the schema, but if you configure the graphql.config.js file, you can also tell the extension to include certain headers (e.g. for auth) and make queries right in vscode
